### DISCUSSION TITLE
[Highlight queries] matching more specific token -- rules when several queries match a token?

### DISCUSSION THREAD

**Discussion by HerrmannM** [tree-sitter/tree-sitter#1267] [Q&A] [answered]
Hi all,

I'm at the point where I'm writing the highlight queries for my parser. The language has various use of parentheses, some that only make sense in a special context. For example, I have the queries
```
[ "(" ")" ","]@delimiter
(term "builtin"@exception  "("@exception  _  ")"@exception  )
```

I wish to have most parentheses colored with `@delimiter`. However, I would like to color the parentheses of "builtin" the same color as "builtin". I would expect it to be the case with the above queries as the parentheses inside 
```
(term "builtin"@exception  "("@exception  _  ")"@exception  )
```
 are more specific (there is more context) than the one in
```
[ "(" ")" ","]@delimiter
```

The work-around I found for now is to put aliases in the parser and use
`
(term "builtin"@exception  "bi("@exception  _  "bi)"@exception  )
`
Is there a way to achieve this without such a workaround? More generally, what happens when several highlight queries matches a node? How is the final color decided?

Thanks!
Cheers,
Mat

**Comment by ahlinc** [comment, marked as answer]
You can try to use Tree-sitter Playground: https://tree-sitter.github.io/tree-sitter/playground
Where for the `C` grammar try to use:

_Source:_
```c
void foo() {
}

typedef struct {
  uint16_t index;
  uint16_t length;
} Bar;
```

_Query:_
```sexp
(function_definition body: (_ [ "{" "}" ] @par.func))
[ "{" "}" ] @par.common
```

<details><summary><b>Screenshot:</b></summary><p>

![image](https://user-images.githubusercontent.com/14666676/125879184-0d434a2a-e2d8-4630-8902-d0a1af94a24f.png)

</p></details>

As you can see on the above screenshot the first matching query have precedence for highlighting over all following queries.
So try to reorder your queries and move more common queries to the bottom of your queries file.

**Reply by HerrmannM** [reply]
Hi,

Thank you for your answer! I try to reverse the queries from your example, and as expected from your explanation, everything was blue. At the moment, I'm mostly testing with neovim, and it seems that some states (e.g. bold) can be "painted over" something already colored.

For example, here, I target the delimiters (parentheses and comma) inside the product (A, A): the blue letters are not bold (but the delimiter are).
```
(type [(ty_name) (ty_fun)]@nonbold)
(type (ty_prod "prod_delim"@bold))
```
![image](https://user-images.githubusercontent.com/2685294/125907376-b361be1e-1d65-40e1-90b3-13b366228b07.png)

If now I target the full product node, not just the delimiters: the letters A keep their color (as you explained), but become bold.
The last 'A' and the arrow, which are not in the product, remain non-bold.
```
(type [(ty_name) (ty_fun)]@nonbold)
(type (ty_prod)@bold)
```
![image](https://user-images.githubusercontent.com/2685294/125907637-aae25bdf-e989-4a02-bb90-81ba7695565d.png)

So more a neovim issue I guess? Unless it is an expected behaviour.
Right now I have trouble testing the highlighter from tree-sitter, as it crashes (https://github.com/tree-sitter/tree-sitter/issues/1232). I'll wait for the next release to see what's the outcome ;-)

Thanks again for your help,
Cheers,
Mat

**Reply by HerrmannM** [reply]
Hi, me again,

I found a case in my code where the color is being overwritten. Note: it may just be a neovim quirk...

I can defined named function like this
![image](https://user-images.githubusercontent.com/2685294/125912806-259f3e03-d214-4aa1-876d-5c4f54f83801.png)
```
(identifier)@identifier
```
the 'f' has the color of the identifier (like the a) - it is bold because it is being painted over by the arrow construction.

Now if I specifically target the identifier like so
```
(identifier)@identifier
(term (lambda (mapsto (identifier)@function)))
```
![image](https://user-images.githubusercontent.com/2685294/125913282-9ec84493-d2c4-4629-b027-ab77b6c77696.png)

Cheers,
Mat

**Reply by ahlinc** [reply]
I supose Neovim uses own implementation for highligting with which I'm not familiar. It's hard to say what's going wrong with queries without the `tree-sitter parse` output.
To test the `tree-sitter highlight` behaviour you can try a previous CLI version by taking it from Github artifacts for this repo.

**Reply by sogaiu** [reply]
I might be off here, but FWIW...

@HerrmannM There was a report recently concerning how neovim handles highlighting that seemed (to me anyway) similar to the situation you described: https://github.com/sogaiu/tree-sitter-clojure/issues/14

IIRC, in the issue, @theHamsta [gave a brief description of neovim's highlighting](https://github.com/sogaiu/tree-sitter-clojure/issues/14#issuecomment-880192734) as:

> The current drawing algorithm is draws smaller ranges over bigger ranges. So as long as the symbols are in the syntax tree and the queries match they will be highlighted.

May be this can help with understanding what is happening.

**Reply by HerrmannM** [reply]
@ahlinc Thanks, I'll try with a previous version soon!

@sogaiu Indeed, looks a lot like what I'm having, I experienced the "smaller ranges over bigger ranges"! it is actually quite an interesting way to paint.

Cheers,
Mat

**Reply by sogaiu** [reply]
@HerrmannM On a side note, may be the following experimental work will be of interest: https://github.com/neovim/neovim/pull/15114

**Reply by HerrmannM** [reply]
@sogaiu this looks promising, I hope it will make it in neovim!
